gen/map.go

```diff
@@ -281,3 +281,147 @@ func (m *mapGenerator) equalsUnhashable(g Generator, spec *compile.MapSpec) (str
 
 	return name, wrapGenerateError(spec.ThriftName(), err)
 }
+
+func (m *mapGenerator) zapMarshaler(
+	g Generator,
+	root *compile.MapSpec,
+	fieldValue string,
+) (string, error) {
+	name := zapperName(g, root)
+	switch compile.RootTypeSpec(root.KeySpec).(type) {
+	case *compile.StringSpec:
+		return m.zapStringKeyMarshaler(g, name, root, fieldValue)
+	default:
+		return m.zapNonstringKeyMarshaler(g, name, root, fieldValue)
+	}
+}
+
+func (m *mapGenerator) zapStringKeyMarshaler(
+	g Generator,
+	name string,
+	root *compile.MapSpec,
+	fieldValue string,
+) (string, error) {
+	if err := g.EnsureDeclared(
+		`
+			<$zapcore := import "go.uber.org/zap/zapcore">
+
+			type <.Name> <typeReference .Type>
+			<$m := newVar "m">
+			<$k := newVar "k">
+			<$v := newVar "v">
+			<$enc := newVar "enc">
+			// MarshalLogObject implements zapcore.ObjectMarshaler, enabling
+			// fast logging of <.Name>.
+			func (<$m> <.Name>) MarshalLogObject(<$enc> <$zapcore>.ObjectEncoder) error {
+				for <$k>, <$v> := range <$m> {
+					<zapEncodeBegin .Type.ValueSpec ->
+						<$enc>.Add<zapEncoder .Type.ValueSpec>((string)(<$k>), <zapMarshaler .Type.ValueSpec $v>)
+					<- zapEncodeEnd .Type.ValueSpec>
+				}
+				return nil
+			}
+			`, struct {
+			Name string
+			Type *compile.MapSpec
+		}{
+			Name: name,
+			Type: root,
+		},
+	); err != nil {
+		return "", err
+	}
```

Comment:
optional nit: can return err with the sprintf. There's no chance of that
panicking because name and fieldValue are strings (non-nillable).

```
err := g.EnsureDeclared(`
  ...
`, ...)
return fmt.Sprintf(...), err
```